placehero/src/avatars.rs

```diff
@@ -23,79 +26,114 @@ struct AvatarResponse {
+                        while let Some(url) = rx.recv().await {
+                            let proxy = proxy.clone();
+                            tokio::task::spawn(async move {
+                                let url = url.avatar_url;
+                                let result = image_from_url(&url).await;
+                                match result {
+                                    // We choose not to handle the case where the event loop has ended
+                                    Ok(image) => drop(proxy.message(AvatarResponse { url, image })),
+                                    // TODO: Report in the frontend?
+                                    Err(err) => {
+                                        tracing::warn!(
+                                            "Loading avatar from {url:?} failed: {err}."
+                                        );
+                                    }
+                                }
+                            });
                         }
-                    });
-                }
-            },
-            |this: &mut Self, tx| {
-                if this.requester.is_some() {
-                    tracing::warn!("Unexpectedly got a second worker for requesting avatars.");
-                }
-                this.requester = Some(tx);
-            },
-            |this: &mut Self, response| {
-                let ret = this.icons.insert(response.url, Some(response.image));
-                if !matches!(ret, Some(None)) {
-                    tracing::warn!("Potentially loaded or tried to load same avatar twice.");
-                }
+                    },
+                    |_: &mut State, this: &mut Self, tx| {
+                        if this.requester.is_some() {
+                            tracing::warn!(
+                                "Unexpectedly got a second worker for requesting avatars."
+                            );
+                        }
+                        this.requester = Some(tx);
+                    },
+                    |_: &mut State, response| response,
+                ),
+            ),
+            // Convert to the user-defined action type by ignoring any action we actually submit.
+            |_, action| match action {
+                MessageResult::Action(_) => MessageResult::RequestRebuild,
+                MessageResult::RequestRebuild => MessageResult::RequestRebuild,
+                MessageResult::Nop => MessageResult::Nop,
+                MessageResult::Stale => MessageResult::Stale,
             },
         )
     }
 
-    pub(crate) fn avatar<State: 'static>(
-        &mut self,
-        url: &str,
-    ) -> impl WidgetView<State> + use<State> {
-        let width = 50.;
-        let height = 50.;
-        if let Some(maybe_image) = self.icons.get(url) {
-            if let Some(image_) = maybe_image {
-                return Either::A(sized_box(image(image_)).width(width).height(height));
+    // TODO: ArcStr for URL?
```

Comment:
(Apart from general documentation) This could probably have a doc comment (stole this from the mason example):

```suggestion
    /// Requires the `Avatars` resource to be [provided](provides).
    // TODO: ArcStr for URL?
```